revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,180 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a certificate chain
+// as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+type Options struct {
+	CertChain       []*x509.Certificate
+	SigningTime     time.Time
+	HTTPClient      *http.Client
+	MergeErrorsFunc func(errors []error) error
+}
+
+const pkixNoCheckOID string = "1.3.6.1.5.5.7.48.1.5"
+const ocspMaxResponseSize int64 = 20480 //bytes
+
+func OCSPStatus(opts Options) error {
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	certResults := make([]error, len(opts.CertChain))
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go certOCSPStatus(cert, opts.CertChain[i+1], &certResults[i], &wg, opts)
+		} else {
+			// Last is root cert, which will never be revoked by OCSP
+			certResults[len(opts.CertChain)-1] = nil
+		}
+	}
+
+	wg.Wait()
+	return opts.MergeErrorsFunc(certResults)
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
```

Comment:
Unfortunately no. I tried using this first, but there is an issue with how the chain is validated with respect to self-signed certificates. For OCSP, we need to ensure that the root is a CA. However, a cert chain with only a root is considered invalid, while a cert chain with only a non-CA self-signed cert is okay. Other than this, the errors that are returned are not very easy to parse, making them difficult to determine what the true issue is and how to return this information best to the user. For all of these reasons, I decided it would be best to just recreate one of the essential functions here to not be tied to the larger validation functions.

Reply:
Checking if each issuer is a CA is a check that is performed when calling the method CheckSignatureFrom (https://cs.opensource.google/go/go/+/refs/tags/go1.20.2:src/crypto/x509/x509.go;l=828-831). This is where the check is performed, where we need to ensure the root is a CA.

If you were referring to the example I'm mentioning above about the behavior for ValidateCodeSigningCertChain, that is a result of the assumption here that a chain with length one is not a CA: https://github.com/notaryproject/notation-core-go/blob/main/x509/cert_validations.go#L45-L58. This is an intended function as described by this test: https://github.com/notaryproject/notation-core-go/blob/main/x509/cert_validations_test.go#L287-L293

Reply:
The cert chain passed to ocsp package is the same cert chain used for signing certificates, correct? If that is the case, it is not clear to why the same cert validations do not work for this chain.

Reply:
The cert chain passed to ocsp package is the same cert chain used for signing certificates, correct? If that is the case, it is not clear to why the same cert validations do not work for this chain.

Reply:
The cert validation does not work because it essentially allows non-CA self-signed certificates to be valid. This is fine for signing, but not for OCSP verification. This is because a root cannot be checked for revocation via OCSP, so ensuring that a self-signed certificate is actually a root is essential. This can be achieved through the use of the [CheckSignatureFrom](https://pkg.go.dev/crypto/x509#Certificate.CheckSignatureFrom) function that checks that a certificate's parent is a valid issuer.

This function is used for chains that are larger than length 1, but not for single certificate chains (which are comprised of only a self-signed certificate). Instead, the assumption is made that the cert must be a leaf and that it cannot be a CA (meaning it cannot be a root).

TLDR: The core issue is the single assumption that [a chain with length 1 must be a non-CA](https://github.com/notaryproject/notation-core-go/blob/main/x509/cert_validations.go#L45-L46). An alternative might be to use this function for chains that are greater in length than a single cert and use my existing functionality otherwise. I can test this to determine if this would work as intended or if there are other unintended consequences

Reply:
The newest commit implements the other function for chains that have a length greater than 1, while maintaining the previous functionality for single certificate chains (as suggested in my previous comment). Resolving this thread as a result

Reply:
For OCSP, we need to ensure that the root is a CA. However, a cert chain with only a root is considered invalid, while a cert chain with only a non-CA self-signed cert is okay. 
> Sorry, I am confused. Your code doesnot validates if the self signed cert is a CA or not(no validation for keyusage). Also, can you please point me to the rfc section which says self signed cert needs to be a CA for OCSP checking?


Other than this, the errors that are returned are not very easy to parse, making them difficult to determine what the true issue is and how to return this information best to the user. For all of these reasons, I decided it would be best to just recreate one of the essential functions here to not be tied to the larger validation functions.
> Can we improve the error returned by the function? It would also help usecase where we are validating codesinging certificates.

Reply:
> Your code does not validates if the self signed cert is a CA or not

This validation is done [here](https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/crypto/x509/x509.go;l=829). It is a part of checking that the issuer is valid. A certificate cannot be issued by a non-CA, so it confirms that it is a CA. This function is [used in our code](https://github.com/notaryproject/notation-core-go/blob/main/x509/cert_validations.go#L114) for any certificate that is not the leaf in a chain longer than length 1. The only issue with our code is how chains of length one are handled.

> can you please point me to the rfc section which says self signed cert needs to be a CA for OCSP checking?

I cannot find any specific mention of this in the RFC, but this is an essential concept of how OCSP works. OCSP is not checked for a root certificate because a root certificate would issue itself, meaning any OCSP server provided by it could not be considered a reliable source of determining if the cert itself has been compromised. This would apply to any self-signed certificate as well. So, it does not make sense to return an okay status for a chain that is comprised of only a non-CA self-signed cert (as there is no way to check for a reliable OCSP response, and it should be verified since is not considered a root since it is not a CA). Rather than just passing with a result of OK for any chain of length 1 with a self-signed certificate, it is given the chain validation error if the cert is not a root, which corresponds to a result of unknown. I believe this is the best approach when considering why OCSP is used.

Thinking about this from a use case perspective, consider a situation where someone is signing using a self-signed certificate. Since there is no way to check the status of this random certificate, should this be a valid revocation check? IMO an Unknown status as a result of a certificate chain error makes the most sense here.

> Can we improve the error returned by the function?

I think this is out of scope for this PR. If we believe this should be improved, we can track it in a new issue.

Reply:
Consensus was to allow any self-signed cert to be used in revocation check (essentially skipped) with a new result of NonRevokable. W.r.t improving error messages, this will be tracked in the following new issue https://github.com/notaryproject/notation-core-go/issues/140